### PR TITLE
Feature/test algos nik

### PR DESCRIPTION
**Pull request by Mr-Cool-in-Git**

### CODE REVIEW

File: Dockerfile
Comment (IcyAltair) [inline Dockerfile:1]: COPY исходников test.py и директории test-data зачем нужны? Ещё 1 слой в образе?
Запуск теста CMD переведён в качестве аргумента запуска контейнера, после MR внесу свои изменения

File: gamac/algorithms/affinity.py [hunk 1]
```diff
@@ -8,109 +8,94 @@
...
+
+    def fit(self, X):
+        X = cp.asarray(X)
+        n_samples = X.shape[0]
+        
+        # Вычисление матрицы схожести (negative squared Euclidean distance)
+        S = -cp.sum((X[:, cp.newaxis, :] - X[cp.newaxis, :, :]) ** 2, axis=2)
+        
+        # Установка предпочтений
         if self.preference is None:
```
Comment (IcyAltair) [inline gamac/algorithms/affinity.py:41]: if not self.preference -?

File: gamac/algorithms/affinity.py [hunk 2]
```diff
@@ -8,109 +8,94 @@
...
-                        second_max = cp.max(AS_row)
-                        R[i, k] = distance_matrix[i, k] - second_max
-                    else:
-                        R[i, k] = distance_matrix[i, k] - max_AS[i]
-
-            for k in range(N):
-                Rp = cp.maximum(R[k], 0)
-                Rp[k] = R[k, k]
-                sum_Rp = cp.sum(Rp) - Rp[k]
-                for i in range(N):
```
Comment (IcyAltair) [inline gamac/algorithms/affinity.py:75]: Вынести инвариант цикла?

File: gamac/algorithms/affinity.py [hunk 3]
```diff
@@ -8,109 +8,94 @@
...
-            labels_=labels_,
-            centroids_=centroids_,
-        )
-
-    def _compute_distance_matrix(self, X):
-        diff = X[:, cp.newaxis, :] - X[cp.newaxis, :, :]
-        return cp.linalg.norm(diff, axis=2)
+            cp.fill_diagonal(S, self.preference)
+        
+        # Инициализация матриц
```
Comment (IcyAltair) [inline gamac/algorithms/affinity.py:46]: Упс... Забыл, что аффинити дропаем

File: gamac/algorithms/agglomerative.py [hunk 4]
```diff
@@ -11,90 +11,86 @@
...
-        """Обновление матрицы расстояний по формуле Ланса-Уильямса"""
-        mask = cp.ones(proximity.shape[0], dtype=bool)
-        mask[[i, j]] = False
-
-        if self.linkage == 'single':
-            proximity[i, mask] = cp.minimum(proximity[i, mask], proximity[j, mask])
+        self.labels_ = None
+
+    def fit(self, X):
+        # X должен быть cupy-массивом
```
Comment (IcyAltair) [inline gamac/algorithms/agglomerative.py:27]: Нужна проверка на тип данных?

File: gamac/algorithms/agglomerative.py [hunk 5]
```diff
@@ -11,90 +11,86 @@
...
+
+        cluster_distances = {}
+        for i in range(n_samples):
+            for j in range(i + 1, n_samples):
+                cluster_distances[(i, j)] = distances[i, j].item()  # .item() для извлечения числа из cupy.float
+
+        while len(clusters) > self.n_clusters:
+            c1, c2 = min(cluster_distances, key=cluster_distances.get)
+            clusters[c1].extend(clusters[c2])
+            del clusters[c2]
```
Comment (IcyAltair) [inline gamac/algorithms/agglomerative.py:40]: Можем попасть по 228 УК РФ? (Index Out of Range Error)?

File: gamac/algorithms/agglomerative.py [hunk 6]
```diff
@@ -11,90 +11,86 @@
...
+        if self.labels_ is None:
+            raise Exception("Model has not been fitted yet.")
+        return self.labels_
+
+    def _compute_distances(self, X):
+        sum_sq = cp.sum(X ** 2, axis=1)
+        distances = cp.sqrt(cp.maximum(sum_sq[:, None] + sum_sq[None, :] - 2 * cp.dot(X, X.T), 0))
+        return distances
+
+    def _linkage_distance(self, X, cluster1, cluster2):
```
Comment (IcyAltair) [inline gamac/algorithms/agglomerative.py:63]: Это свой метод вычисления расстояния, не коммунальный на все алгосы?

File: gamac/algorithms/birch.py [hunk 7]
```diff
@@ -1,214 +1,159 @@
...
-        new_node2.parent = self.parent
-
-        # Шаг 3: Распределяем подкластеры между новыми узлами
-        for cf in self.subclusters:
-            dist1 = cp.linalg.norm(cf.centroid - self.subclusters[idx1].centroid)
-            dist2 = cp.linalg.norm(cf.centroid - self.subclusters[idx2].centroid)
-            
-            if dist1 < dist2:
-                new_node1.subclusters.append(cf)
+                if distances[closest] <= self.threshold:
```
Comment (IcyAltair) [inline gamac/algorithms/birch.py:78]: Вот тут точно ли случай < и случай = уйдут в 1 и тот же кейс?

File: gamac/algorithms/birch.py [hunk 8]
```diff
@@ -1,214 +1,159 @@
...
-            self.parent = new_root
-
-        # Шаг 5: Проверяем необходимость рекурсивного разделения
-        for node in [new_node1, new_node2]:
-            if len(node.subclusters) > self.branching_factor:
-                node._split()
+                self.cfs.append(cf)
 
+    def find_closest_child(self, data):
+        centroids = [cf.centroid().get() for cf in self.cfs]  # Move to CPU for compatibility
```
Comment (IcyAltair) [inline gamac/algorithms/birch.py:88]: Тут бы проверок добавить, что мы валидно память выделяем

File: gamac/algorithms/birch.py [hunk 9]
```diff
@@ -1,214 +1,159 @@
...
+            point_cpu = point.get()
+            closest = np.argmin(np.linalg.norm(subcluster_centers - point_cpu, axis=1))
+            self.labels_.append(self.subcluster_labels[closest])
+        self.labels_ = cp.array(self.labels_, dtype=cp.int32)
+
+        return BirchModel(labels_=self.labels_, subcluster_labels=self.subcluster_labels, tree=self.tree)
+
+    def predict(self, X):
+        X_gpu = cp.array(X)
+        subclusters = np.array([cf.centroid().get() for cf in self.tree.root.cfs])
```
Comment (IcyAltair) [inline gamac/algorithms/birch.py]: np or cp? Вот в чём вопрос

File: gamac/algorithms/dbscan.py [hunk 10]
```diff
@@ -1,80 +1,159 @@
...
+        super().__init__(labels_)
+        self.X_ = X_
+        self.eps_sq = eps ** 2
+
+    def predict(self, df: DataFrameType) -> LabelsType:
+        df = cp.asarray(df, dtype=cp.float32)
+        X_train = self.X_
+        
+        # Батчевый расчет расстояний для больших данных
+        batch_size = 4096
```
Comment (IcyAltair) [inline gamac/algorithms/dbscan.py:21]: Вынести в конфиг? Задать на уровне родительского класса?

File: gamac/algorithms/dbscan.py [hunk 11]
```diff
@@ -1,80 +1,159 @@
...
+            queue = queue[1:]
+            cluster.append(int(current))
+            
+            neighbors = self._get_neighbors(current)
+            mask = ~visited[neighbors]
+            new_points = neighbors[mask]
+            
+            # Пометка core-точек заранее
+            core_mask = is_core[new_points]
+            queue = cp.concatenate((queue, new_points[core_mask]))
```
Comment (IcyAltair) [inline gamac/algorithms/dbscan.py:83]: Вот тут проверки на размерность не нужны?

File: gamac/algorithms/hdbscan.py [hunk 12]
```diff
@@ -1,225 +1,92 @@
...
 from collections import defaultdict
 
+from gamac.algorithms.base import ClusteringModel, ClusteringAlgo, AlgoConfig
+from gamac.data.data_pipeline import DataFrameType, LabelsType
 
-class HDBSCAN:
-    def __init__(self, min_cluster_size=5, min_samples=None, metric='euclidean',
-                 alpha=1, cluster_selection_epsilon=0.1,
-                 cluster_selection_method='eom', allow_single_cluster=False):
+pylibraft.config.set_output_as("cupy")
```
Comment (IcyAltair) [inline gamac/algorithms/hdbscan.py:8]: Внезапно

File: gamac/algorithms/hdbscan.py [hunk 13]
```diff
@@ -1,225 +1,92 @@
...
 
     def predict(self, X):
-        """Predict using nearest core points"""
-        X_gpu = cp.asarray(X, dtype=cp.float32)
-        dists = self._pairwise_distance(X_gpu)
-        nearest = cp.argmin(dists, axis=1)
-        return self.labels_[nearest.get()]
\ No newline at end of file
+        if not hasattr(self, 'labels_'):
+            raise Exception("Model not fitted yet")
```
Comment (IcyAltair) [inline gamac/algorithms/hdbscan.py]: Raise GPU контекста? 2 в журнал и дневник, вызов родителей в школу

File: gamac/algorithms/kmeans.py [hunk 14]
```diff
@@ -23,84 +34,90 @@ class KMeans:
...
+            random_state=None
+    ):
+        super().__init__()
+        self.n_clusters = n_clusters
+        self.max_iter = max_iter
+        self.tol = tol
+        self.random_state = random_state
+        self.centroids = None
+
+    def fit(self, X):
```
Comment (IcyAltair) [inline gamac/algorithms/kmeans.py:54]: Press F for C вставки

File: gamac/algorithms/kmeans.py [hunk 15]
```diff
@@ -23,84 +34,90 @@ class KMeans:
...
+                cluster_points = X[labels == i]
+                if cluster_points.shape[0] > 0:
+                    new_centroids[i] = cluster_points.mean(axis=0)
+                else:
+                    # Если кластер пуст, сохраняем прежний центроид
+                    new_centroids[i] = self.centroids[i]
+
+            # Проверка условия сходимости (изменение центроидов)
+            centroid_shift = cp.max(cp.linalg.norm(new_centroids - self.centroids, axis=1))
+            if centroid_shift <= self.tol:
```
Comment (IcyAltair) [inline gamac/algorithms/kmeans.py:88]: Точно <=? Не <?

File: gamac/algorithms/meanshift.py [hunk 16]
```diff
@@ -1,34 +1,95 @@
...
+        X = cp.asarray(X)  # Конвертация в CuPy массив
+        centroids = X.copy()
+        
+        for _ in range(self.max_iter):
+            max_shift = 0.0
+            for i in range(len(centroids)):
+                centroid = centroids[i]
+                distances = cp.linalg.norm(X - centroid, axis=1)
+                in_window = distances <= self.bandwidth
+                if not cp.any(in_window):
```
Comment (IcyAltair) [inline gamac/algorithms/meanshift.py:44]: Кажется здесь надо вывести контекст инфу в логгировании, да и вообще что-то logger не присутствует(

File: gamac/autoclustering.py [hunk 17]
```diff
@@ -39,9 +49,9 @@ def _check_input(
             text: Optional[list[str]],
             image: Optional[list[Image]],
     ):
-        if not table and not text and not image:
+        if table is None and text is None and image is None:
```
Comment (IcyAltair) [inline gamac/autoclustering.py]: Триплет проверку на None поставить сразу?

File: gamac/estimation/functions.py [hunk 18]
```diff
@@ -37,8 +37,11 @@ def br(container: EstimationContainer) -> float:
         cl_var = (cl_d * cl_d).mean().item()
         var_stable = 1.0 + cl_var + 1e-6
         result_acc += cl_n * np.log(var_stable)
-    result = result_acc / container.n
-    return -result
+    if container.n:
+        result = result_acc / container.n
+        print(result_acc, container.n)
+        return -result
+    return -100000
```
Comment (IcyAltair) [inline gamac/estimation/functions.py:44]: Вот тут не понял, -100000 вычислено эмпирически?

File: test.py [hunk 19]
```diff
@@ -7,10 +7,11 @@
 
 
 def main():
-    data = pd.read_csv('test-data/gen.csv').head(1000).values
```
Comment (IcyAltair) [inline test.py:10]: Переделаю с argparser

**Review by IcyAltair — COMMENTED** [review]
10 сишных вставок из 10

File: Dockerfile
Comment (Mr-Cool-in-Git) [reply Dockerfile:1]: Ок, это когда я тестил с винды, нужно было новые COPY указать, чтобы в контейнере было видно
Потом перешел на арендованный сервак с убунту и запускал напрямую

File: gamac/algorithms/affinity.py (hunk 1 above)
Comment (Mr-Cool-in-Git) [reply gamac/algorithms/affinity.py:41]: По PEPу указывается таким образом явно. 
Также not var не со всеми объектами нормально работает

File: gamac/algorithms/affinity.py (hunk 3 above)
Comment (Mr-Cool-in-Git) [reply gamac/algorithms/affinity.py:46]: Надо оставить, чтобы протестить на других датасетах
на gen.csv выдавало нули в метках

File: gamac/algorithms/agglomerative.py (hunk 4 above)
Comment (Mr-Cool-in-Git) [reply gamac/algorithms/agglomerative.py:27]: по идее нет, в проекте заранее массив оборачивается в cp.array

File: gamac/algorithms/kmeans.py (hunk 15 above)
Comment (Mr-Cool-in-Git) [reply gamac/algorithms/kmeans.py:88]: Тут без разницы имхо с учетом порядков чисел

File: gamac/algorithms/agglomerative.py (hunk 5 above)
Comment (Mr-Cool-in-Git) [reply gamac/algorithms/agglomerative.py:40]: По идее нет

Comment (Mr-Cool-in-Git) [reply gamac/algorithms/agglomerative.py:40]: но все возможно, агломеративка нерабочая (в плане она зависает и не выполняется по времени)

File: gamac/estimation/functions.py (hunk 18 above)
Comment (Mr-Cool-in-Git) [reply gamac/estimation/functions.py:44]: Это константа-штраф при наличии только аномальных меток для оптуны. Исходил от того, в какую сторону оптуна тюнит метрику

File: test.py (hunk 19 above)
Comment (Mr-Cool-in-Git) [reply test.py:10]: кк

File: gamac/autoclustering.py (hunk 17 above)
Comment (Mr-Cool-in-Git) [reply gamac/autoclustering.py]: добавил

**Review by IcyAltair — APPROVED** [review]
10/10 слитых CS mm каток

File: gamac/algorithms/agglomerative.py [hunk 20]
```diff
@@ -11,90 +11,86 @@
...
+            for key in keys_to_remove:
+                del cluster_distances[key]
+
+            for c in clusters:
+                if c != c1:
+                    dist = self._linkage_distance(X, clusters[c1], clusters[c])
+                    key = (min(c1, c), max(c1, c))
+                    cluster_distances[key] = dist
+
+        self.labels_ = cp.empty(n_samples, dtype=cp.int32)
```
Comment (parallaxel) [inline gamac/algorithms/agglomerative.py]: You should not assign labels to CLusteringAlgo instance
You should use ClusteringModel to hold labels

File: gamac/algorithms/agglomerative.py [hunk 21]
```diff
@@ -11,90 +11,86 @@
...
+        for cluster_id, points in enumerate(clusters.values()):
+            for point in points:
+                self.labels_[point] = cluster_id
+        return AgglomerativeClusteringModel(labels_=self.labels_)
+
+    def predict(self, X):
```
Comment (parallaxel) [inline gamac/algorithms/agglomerative.py]: You do not need this method in ClusteringAlgo
ClusteringModel implements it

File: gamac/algorithms/birch.py [hunk 22]
```diff
@@ -1,214 +1,159 @@
...
+
+        # Проверка количества подкластеров
+        if len(subclusters) < 2:
+            self.subcluster_labels = np.zeros(len(subclusters), dtype=int)
+            self.labels_ = cp.zeros(len(X), dtype=cp.int32)
+            return BirchModel(labels_=self.labels_, subcluster_labels=self.subcluster_labels, tree=self.tree)
+
+        # Адаптация числа кластеров
+        n_clusters = min(self.n_clusters, len(subclusters))
+        clustering = AgglomerativeClustering(n_clusters=n_clusters)
```
Comment (parallaxel) [inline gamac/algorithms/birch.py]: You should use lightweight algorithm here, as Birch is proved to be of linear complexity

File: gamac/algorithms/meanshift.py [hunk 23]
```diff
@@ -1,34 +1,95 @@
...
+            if cp.min(distances) > self.bandwidth:
+                unique_centroids.append(centroid)
+        self.centroids = cp.array(unique_centroids)
+        
+        # Назначение меток
+        self.labels = self._assign_labels(X)
+        return MeanShiftModel(labels_=self.labels, centroids_=self.centroids)
+    
+    def _assign_labels(self, X):
+        labels = cp.zeros(X.shape[0], dtype=cp.int32)
```
Comment (parallaxel) [inline gamac/algorithms/meanshift.py]: Better be cp.empty to prevent unnecessary filling with zeros

File: gamac/algorithms/birch.py [hunk 24]
```diff
@@ -1,214 +1,159 @@
 import cupy as cp
+import numpy as np
 import pylibraft.config
+from sklearn.cluster import AgglomerativeClustering
```
Comment (parallaxel) [inline gamac/algorithms/birch.py:4]: ...

File: gamac/algorithms/agglomerative.py (hunk 21 above)
Comment (Mr-Cool-in-Git) [reply gamac/algorithms/agglomerative.py]: Done

File: gamac/algorithms/agglomerative.py (hunk 20 above)
Comment (Mr-Cool-in-Git) [reply gamac/algorithms/agglomerative.py]: Done

File: gamac/algorithms/meanshift.py (hunk 23 above)
Comment (Mr-Cool-in-Git) [reply gamac/algorithms/meanshift.py]: Done

File: gamac/algorithms/birch.py (hunk 22 above)
Comment (Mr-Cool-in-Git) [reply gamac/algorithms/birch.py]: Done
